presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RewriteExpressionWithConstantVariable.java

```diff
@@ -0,0 +1,462 @@
+import com.facebook.presto.sql.planner.plan.OffsetNode;
+import com.facebook.presto.sql.planner.plan.SampleNode;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+import com.facebook.presto.sql.planner.plan.SortNode;
+import com.facebook.presto.sql.planner.plan.UnnestNode;
+import com.facebook.presto.sql.planner.plan.WindowNode;
+import com.facebook.presto.sql.relational.FunctionResolution;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.SystemSessionProperties.isRewriteExpressionWithConstantEnabled;
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.common.type.DateType.DATE;
+import static com.facebook.presto.common.type.IntegerType.INTEGER;
+import static com.facebook.presto.common.type.VarcharType.VARCHAR;
+import static com.facebook.presto.expressions.LogicalRowExpressions.extractConjuncts;
+import static com.facebook.presto.spi.plan.AggregationNode.singleGroupingSet;
+import static com.facebook.presto.sql.planner.plan.ChildReplacer.replaceChildren;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+
+/**
+ * Get constant from filter and project node. Rewrite expressions in parent nodes with the constant.
+ * For example, for query "select orderkey, orderpriority, avg(totalprice) from orders where orderpriority='3-MEDIUM' group by 1, 2", the query plan changes from
+ * <pre>
+ *     - OutputNode
+ *          orderkey, orderpriority, avg
+ *          - Aggregate
+ *              avg := avg(totalprice)
+ *              Grouping Keys := [orderkey, orderpriority]
+ *              - Filter
+ *                  orderpriority = '3-MEDIUM'
+ *                  - TableScan
+ *                      orderkey := orderkey
+ *                      orderpriority := orderpriority
+ *                      totalprice := totalprice
+ * </pre>
+ * to
+ * <pre>
+ *      - OutputNode
+ *          orderkey, expr_12, avg
+ *          - project
+ *              orderkey := orderkey
+ *              expr_12 := '3-MEDIUM'
+ *              avg := avg
+ *              - Aggregate
+ *                  avg := avg(totalprice)
+ *                  Grouping Keys := [orderkey]
+ *                  - Filter
+ *                      orderpriority = '3-MEDIUM'
+ *                      - TableScan
+ *                          orderkey := orderkey
+ *                          orderpriority := orderpriority
+ *                          totalprice := totalprice
+ * </pre>
+ */
+public class RewriteExpressionWithConstantVariable
```

Comment:
I think I'm missing something very basic, but I'm still pretty confused about how this optimizer works. I *think* basically what it does, is for a handful of cases, it will replace some known constant variable references with the constant itself.  And then maybe something else like PruneUnreferencedOutputs prunes usages below that of those variables?  I still don't understand how the final project node gets added though unless there was already a project node there.  I think this optimizer would benefit from plan tests that exercise just this optimizer so it would be easier to see what this rewrite is doing by itself.

Reply:
The benefit of this optimizer is in three aspects: 
- add a projection to assign a variable with a constant when it's constant. The benefit is that, at runtime, constant projection assignments return RunLengthEncoded block, which consumes less memory, and we also have a handful of runtime optimizations for RunLengthEncoded block 
- reduce the amount of data propagated during query execution. For example, if we have a filter on a column at the very bottom of the query and this column is not needed in some intermediate operators, we can avoid passing this column between these operators. A simplified example is `select orderkey, l.partkey, o.orderstatus from lineitem l join orders o using (orderkey) where orderstatus='F'`, which the optimization, `orderstatus` will not be in the join hashtable.
- Simplify some nodes, for example, Sort on constant variables, group by constant variables etc.

> I still don't understand how the final project node gets added though unless there was already a project node there

There are project nodes during optimization, they are pruned during optimizations by the pruning rules.